docs/unit-testing.md

```diff
@@ -0,0 +1,1045 @@
+# Unit Testing Guidelines
+
+## Use Jest
+
+[Jest](https://jestjs.io/) is the testing framework of choice for unit tests at MetaMask. It offers a more comprehensive set of features out of the box over Mocha or Tape that make writing tests for production code very easy. Some of the useful features are:
+
+- [Module mocks](https://jestjs.io/docs/next/mock-functions#mocking-modules)
+- [Timer mocks](https://jestjs.io/docs/next/timer-mocks)
+- [Snapshots](https://jestjs.io/docs/next/snapshot-testing)
+- Automatic [parallelization of tests](https://jestjs.io/blog/2016/09/01/jest-15#buffered-console-messages) without [restricting usage of exclusive tests](https://mochajs.org/#exclusive-tests-are-disallowed)
+- A [consistent](https://jestjs.io/blog/2016/03/11/javascript-unit-testing-performance#adding-everything-up) [emphasis](https://jestjs.io/blog/2016/09/01/jest-15#new-cli-error-messages-and-summaries) on [great developer experience](https://jestjs.io/blog/2017/01/30/a-great-developer-experience)
+
+## Colocate test files with implementation files
+
+Test files are much easier to find when they sit next to the code they test.
+
+🚫
+
+```
+src/
+  permission-controller.ts
+test/
+  permission-controller.ts
+```
+
+✅
+
+```
+src/
+  permission-controller.ts
+  permission-controller.test.ts
+```
+
+## Use `describe` to group tests for the module under test
+
+Wrap all tests for a file in a `describe` block named the same as the file itself:
```

Comment:
I'm curious why this is necessary? Jest will print the name of the file regardless.

Reply:
Does it? I know it prints the name of the file where tests are currently running, but for the report I thought they all got combined

Reply:
Oh I see, no you're right, it always shows the filename in the console output. I'm not sure why we have this rule then.

I guess we need a `describe` block often enough for hooks, and it's nice to keep everything consistent. But maybe the strict requirement for having just one is a relic from when we used mocha.

Reply:
> I guess we need a `describe` block often enough for hooks, and it's nice to keep everything consistent.

Right, that was exactly my thought. However, you're right, this isn't necessary, especially if we are already using `describe` to group tests for the same function/method. I can remove this.

Reply:
Done here: [34f61b6](https://github.com/MetaMask/contributor-docs/pull/38/commits/34f61b6f7913ac313b0772a3c21a34e1b7111e4e)